### ISSUE TITLE
Render before StateChanged

### ISSUE THREAD

**Issue by deadtoadroad** [mrpmorris/Fluxor#62]
I would like to copy something from state to a local property so I can bind to it without modifying state directly. A later event will send the local copy back to state via an action.

If I use `IState.StateChanged` to do this, I sometimes get a render before the `StateChanged` event is fired. This makes mixing state with local properties problematic as it can produce scenarios not covered by tests.

For example, in the page below, the last paragraph should never show, but occasionally does.

    @page "/customers/{id:int}"
    @using Fluxor
    @using FluxorGap.Data
    @using FluxorGap.Store.EditCustomer
    @inherits Fluxor.Blazor.Web.Components.FluxorComponent

    @if (State.Value.IsLoading) {
        <p><em>Loading...</em></p>
    }
    else if (Customer != null) {
        <p>Customer form here</p>
    }
    else {
        <p style="color: red;"><strong>THIS SHOULD NEVER SHOW</strong></p>
    }

    @code {
        [Parameter]
        public int Id { get; set; }

        [Inject]
        private IState<State> State { get; set; }

        [Inject]
        private IDispatcher Dispatcher { get; set; }

        private Customer Customer { get; set; }

        protected override void OnInitialized()
        {
            base.OnInitialized();

            State.StateChanged += StateOnStateChanged;

            Dispatcher.Dispatch(new GetCustomerAction(Id));
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
                State.StateChanged -= StateOnStateChanged;
            
            base.Dispose(disposing);
        }

        private void StateOnStateChanged(object sender, State state)
        {
            Customer = state.Customer?.Clone();
            StateHasChanged();
        }
    }

Is there another way to copy state immediately after an effect?

[fluxor-gap.zip](https://github.com/mrpmorris/Fluxor/files/5020785/fluxor-gap.zip)

**Comment by mrpmorris** [comment]
Have you tried adding your subscription before calling `base.OnInitialized`?

**Comment by deadtoadroad** [comment]
Is this what you mean? Unfortunately there's no change, the red text still occasionally flashes on the screen.

    protected override void OnInitialized()
    {
        State.StateChanged += StateOnStateChanged;

        base.OnInitialized();

        Dispatcher.Dispatch(new GetCustomerAction(Id));
    }

**Comment by mrpmorris** [comment]
Can you show me exactly what you are doing? Both of the events are triggered synchronously so order shouldn't matter.

**Comment by deadtoadroad** [comment]
Hi @mrpmorris, thanks for your time.

The [attached project](https://github.com/mrpmorris/Fluxor/files/5020785/fluxor-gap.zip) (also [here](https://github.com/deadtoadroad/fluxor-gap)) shows one scenario where I encounter this bug. If you open the project and flip between the customers page and any other page a dozen or so times, you will see red text appear briefly once or twice. That's the bug.

As mentioned earlier I would like to copy some of the state locally, specifically the customer object, so that I can bind to it without modifying state directly. In this case it would be a form where input components bind directly to customer properties. Once the user makes their changes, I dispatch the local customer object in an update action that attempts to validate and update the customer via an effect calling a service. Validation errors or a newly updated customer object come back to state.

My colleague has made a PR (#64) that solves the problem. There's another PR (#65) that solves the problem too and I think gives easier access to changed state without having to manage the creation and disposal of event handlers. You can see what effect it would have on the example project [here](https://github.com/deadtoadroad/fluxor-gap/commit/12a13981715a0ef846c26559d09a18476bfcd265#diff-0d3e93e0f6a42a92b3e6c3dcb35729bd).

**Comment by mrpmorris** [comment]
Would you like an extension that notifies you whenever an action is dispatched and lets you grab data out of it without having to reduce it into state at all?

**Comment by deadtoadroad** [comment]
@mrpmorris That's an interesting idea. However, I think access to the entire state might be safer. Tests on actions, reducers and state would cover all the scenarios a page/component encounters. If a page hooks into an action, I'm not sure how that could be as easily tested.

Another approach I thought might solve the issue is an asynchronous `Dispatch` method, one that I could await until all actions and effects complete. Then I could copy state immediately after the await. Is that an option?

**Comment by mrpmorris** [comment]
Async dispatch is not something I would consider including.

**Comment by 82amp** [comment]
> show me exactly what you are doing? Both of the events are triggered synchronously so order shouldn't m

`untypedStateChanged?.Invoke(this, EventArgs.Empty); `  amongst other things, invokes a re-render of the page 
`stateChanged?.Invoke(this, newState);` invokes event handlers that a dev may have added

issue: if the component render fragment uses mixed state and local variables, there is a point at which the page is rendered, but the local copy of state is not yet updated. In this case, for the fluxor-gap example, it is flashing/flickering the red text very briefly.

nice to fix: devs needs to manually call StateHasChanged() in the eventhandler, devs having to manage StateChanged in initialization and disposal 

By re-ordering the calls, we can ensure that the component's state has been updated with fluxor state prior to re-rendering.

There are a couple of other options I can think of: 

1. As per @deadtoadroad's fix, put the subscriber notification into the FluxorComponent's StateSubscription handler (prior to invoking StateHsChanged). The nicety here is no need to subscribe/unsubscribe, it can be managed with virtual/overrides. It keeps the rendering managed by the Component (rather than being driven by the feature). 

2. use the same mechanism that devs use to add an event handler to add the StateSubscription, and process the events in the order they were added. This gives devs the option to add eventhandlers for before or after the overall StateHasChanged, (however, it does not remove the need for un/subscribing)

3. Disable ShouldRender() whilst all event processing occurs (however, this is much more likely to introduce delays on the UI, so I hesitate even to mention it)

**Comment by mrpmorris** [comment]
Merged into release/3.3

Thanks!

**Comment by mrpmorris** [comment]
Note that you can now intercept actions in your UI, so there is no need to store mutable objects (DTOs) in your immutable store.

```
protected override void OnInitialized()
{
  base.OnInitialized();
  SubscribeToAction<ActionThatContainsTheDto>(x => this.Model = x.Model);
}
```
